pkg/xds/server/resource_warming_forcer.go

```diff
@@ -0,0 +1,124 @@
+//   6) Envoy sends DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"] // confirmations that it received endpoints (ACK)
+//
+//   Then if we send a Cluster update (continuing the flow above)
+//   7) Kuma sends DiscoveryResponse [type=CDS, version="UUID-2", nonce="3"] // response with cluster update
+//   8) Envoy sends DiscoveryRequest [type=CDS, version="UUID-2", nonce="3"] // ACK
+//   9) Envoy sends DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"] // Envoy sends a request which looks like the second ACK for the previous endpoints
+//
+//   Updated Cluster is now in warming state until we send DiscoveryResponse with EDS.
+//   We could sent the same DiscoveryResponse again: DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"], but there is no API in go-control-plane to do it.
+//   Instead we set a new version of the Endpoints to force a new EDS exchange:
+//   10) Kuma sends DiscoveryResponse [type=EDS, version="UUID-3", nonce="3"] // triggered because we set snapshot with a new version
+//   11) Envoy sends DiscoveryRequest [type=EDS, version="UUID-3", nonce="3"] // ACK
+//   After this exchange, cluster is now out of the warming state.
+//
+// The same problem is with Listeners and Routes (change of the Listener that uses RDS requires RDS DiscoveryResponse), but since we don't use RDS now, the implementation is for EDS only.
+// More reading of how Envoy is trying to solve it https://github.com/envoyproxy/envoy/issues/13009
+type resourceWarmingForcer struct {
+	util_xds.NoopCallbacks
+	cache  envoy_cache.SnapshotCache
+	hasher envoy_cache.NodeHash
+
+	sync.Mutex
+	lastEndpointNonces map[xds.StreamID]string
+	nodeIDs            map[xds.StreamID]string
+}
+
+func newResourceWarmingForcer(cache envoy_cache.SnapshotCache, hasher envoy_cache.NodeHash) *resourceWarmingForcer {
+	return &resourceWarmingForcer{
+		cache:              cache,
+		hasher:             hasher,
+		lastEndpointNonces: map[xds.StreamID]string{},
+		nodeIDs:            map[xds.StreamID]string{},
+	}
+}
+
+func (r *resourceWarmingForcer) OnStreamClosed(streamID int64) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.lastEndpointNonces, streamID)
+	delete(r.nodeIDs, streamID)
+}
+
+func (r *resourceWarmingForcer) OnStreamRequest(streamID xds.StreamID, request *envoy_api.DiscoveryRequest) error {
+	if request.TypeUrl != envoy_resource.EndpointType {
+		return nil // we force Cluster warming only on receiving the same EDS Discovery Request
+	}
+	if request.ResponseNonce == "" {
+		return nil // initial request, no need to force warming
+	}
+
+	r.Lock()
+	lastEndpointNonce := r.lastEndpointNonces[streamID]
+	r.lastEndpointNonces[streamID] = request.ResponseNonce
+	nodeID := r.nodeIDs[streamID]
+	if nodeID == "" {
+		nodeID = r.hasher.ID(request.Node) // request.Node can be set only on first request therefore we need to save it
+		r.nodeIDs[streamID] = nodeID
+	}
+	r.Unlock()
+
+	if lastEndpointNonce == request.ResponseNonce {
```

Comment:
If I get it right this situation might also happen if CP receives NACK from Envoy. In order to distinguish warming and NACK I guess we have to check `error_detail` as well - https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#nack. What do you think?

Reply:
good catch! I'll do extra check on NACK